src/include/catalog/catalog.h

```diff
@@ -0,0 +1,325 @@
+
+  /**
+   * Get a pointer to the storage table, by table_name.
+   * Supports both catalog tables, and user created tables.
+   *
+   * @param db_oid database that owns the table
+   * @param table_name returns the storage table point for this table
+   * @return a pointer to the catalog
+   * @throw out_of_range exception if either oid doesn't exist or the catalog doesn't exist.
+   */
+  std::shared_ptr<catalog::SqlTableRW> GetCatalogTable(db_oid_t db_oid, const std::string &table_name);
+
+  /**
+   * The global counter for getting next oid. The return result should be converted into corresponding oid type
+   *
+   * This function is atomic.
+   *
+   * @return uint32_t the next oid available
+   */
+  uint32_t GetNextOid();
+
+  /**
+   * TODO(pakhtar): Implement shutdown.
+   */
+
+  /*
+   * Destructor
+   */
+  ~Catalog() = default;
+
+  //  ~Catalog() {
+  //    // destroy all DB
+  //    // DestroyDB(DEFAULT_DATABASE_OID);
+  //  }
+
+  // methods for catalog initializations
+  /**
+   * Add a catalog to the catalog mapping
+   * @param db_oid database oid
+   * @param table_oid table oid
+   * @param name of the catalog
+   * @param table_rw_p catalog storage table
+   */
+  void AddToMaps(db_oid_t db_oid, table_oid_t table_oid, const std::string &name,
+                 std::shared_ptr<SqlTableRW> table_rw_p) {
+    map_[db_oid][table_oid] = std::move(table_rw_p);
+    name_map_[db_oid][name] = table_oid;
+  }
+
+  /**
+   * Utility function for adding columns in a table to pg_attribute. To use this function, pg_attribute has to exist.
+   * @param txn the transaction that's adding the columns
+   * @param db_oid the database the pg_attribute belongs to
+   * @param table the table which the columns belong to
+   */
+  void AddColumnsToPGAttribute(transaction::TransactionContext *txn, db_oid_t db_oid,
+                               const std::shared_ptr<storage::SqlTable> &table);
+
+  /**
+   * Set values for unused columns.
+   * @param vec append to this vector of values
+   * @param cols vector of column types
+   */
+  void SetUnusedColumns(std::vector<type::TransientValue> *vec, const std::vector<SchemaCol> &cols);
+
+  /**
+   * Convert type id to schema type
+   * @param type_id type id
+   * @return schema type
+   */
+  type::TransientValue ValueTypeIdToSchemaType(type::TypeId type_id);
+
+  /**
+   * -------------
+   * Debug support
+   * -------------
+   */
+
+  void Dump(transaction::TransactionContext *txn, db_oid_t db_oid);
+
+ private:
+  /**
+   * Add a row into pg_database
+   */
+  void AddEntryToPGDatabase(transaction::TransactionContext *txn, db_oid_t oid, const std::string &name);
+
+  /**
+   * Add columns created for Postgres compatibility, but unused, to the schema
+   * @param db_p - shared_ptr to database
+   * @param cols - vector specifying the columns
+   *
+   */
+  void AddUnusedSchemaColumns(const std::shared_ptr<catalog::SqlTableRW> &db_p, const std::vector<SchemaCol> &cols);
```

Comment:
I don't understand the purpose of specifying which columns are unused? This sounds like a premature optimization it is going to be a pain to maintain in the future.

Reply:
The reasons why the Unused columns definition was added:

- To make it clearer which columns matter. Prior to this, the catalog code when initializing, sometimes had a few real columns initialized, then many unused columns initialized. This made it hard to see what mattered.
- Unused VARCHAR columns MUST be initialized, otherwise reading from the table generates segmentation faults. This was a recurring problem in earlier instances of the code. Scalar columns can be left unset, but will then contain garbage. Having a specification of which columns are unused allows them to be initialized with a single call.

Reply:
Merged used and unused columns.